### ISSUE TITLE
[0.22] `iwyu_include_picker.cc:1359`: Assertion failed: (*map)[key] == visibility Same file seen with two different visibilities: <debug/safe_iterator.h> Old vis: 2 New vis: 1

### ISSUE THREAD

**Issue by Jacobfaib** [include-what-you-use/include-what-you-use#1490]
To reproduce:
```
$ include-what-you-use -Xiwyu --mapping_file=/path/to/include-what-you-use/0.22/libexec/share/include-what-you-use/gcc.stl.headers.imp
```

If I had to guess, these lines are out of sync:
- https://github.com/include-what-you-use/include-what-you-use/blob/master/gcc.stl.headers.imp#L311 (`{ include: ["<debug/safe_iterator.tcc>", private, "<debug/safe_iterator.h>", public ] },`)
- https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_include_picker.cc#L806 (`{ "<debug/safe_iterator.tcc>", kPrivate, "<debug/safe_iterator.h>", kPrivate },`)

**Comment by kimgr** [comment]
Oops! I think it's actually https://github.com/include-what-you-use/include-what-you-use/blob/78472a725b1dd414e0c6ccdef464a8058593d6eb/gcc.stl.headers.imp#L165 vs. https://github.com/include-what-you-use/include-what-you-use/blob/78472a725b1dd414e0c6ccdef464a8058593d6eb/gcc.stl.headers.imp#L313

I had no idea that safe_iterator.tcc mapping existed since 2012.

Just FYI, many of the `.imp` files shipped with IWYU are also hard-coded inside IWYU, so you typically can't pass these mapping files like this without conflicts. They're intended for use together with `-Xiwyu --no_default_mappings`, which suppresses the hard-coded ones (and lets you pick and choose among the external mappings).

That said, it's reproducible with 
```
include-what-you-use -Xiwyu --no_default_mappings -Xiwyu --mapping_file=iwyu.gcc.imp t.cc
```
too, and works when the extra mapping is removed.

**Comment by Jacobfaib** [comment]
> Just FYI, many of the .imp files shipped with IWYU are also hard-coded inside IWYU, so you typically can't pass these mapping files like this without conflicts. 

Ah, I did not know this! I found them when I had to start rolling my own mapping file for Apple clang's STL (I'd be happy to share that too if it helps).

BTW, by "many" do you mean all of them? Or a subset? And if a subset, is it clearly defined or just whatever you happened to include?

**Comment by kimgr** [comment]
Apples libc++ should already be covered by built-in mappings in latest IWYU (but there may be omissions, of course).

`include_picker.cc` has a number of global tables, and I believe they're all mirrored as `.imp` files. I've checked the ones I think are dual-maintained:
```
[ ] boost-1.64-all.imp
[ ] boost-1.64-all-private.imp
[ ] boost-1.75-all.imp
[ ] boost-1.75-all-private.imp
[ ] boost-all.imp
[ ] boost-all-private.imp
[ ] clang-6.intrinsics.imp
[ ] gcc-8.intrinsics.imp
[x] gcc.libc.imp
[x] gcc.stl.headers.imp
[x] gcc.symbols.imp
[ ] iwyu.gcc.imp   # but this one is actually a "compilation" of all built-in GCC headers
[x] libcxx.imp
[ ] python2.7.imp
[ ] python3.8.imp
[ ] qt4.imp
[ ] qt5_11.imp
[ ] qt5_4.imp
[x] stl.c.headers.imp
[x] stl.public.imp
```

The reason we keep them in two places is because sometimes it's necessary to use `--no_default_mappings` either to get a baremetal environment (e.g. compiling an OS kernel) or to work around bugs, but it might still be useful to add a subset of the mappings or adjust them slightly for your environment and add them explicitly.

**Comment by kimgr** [comment]
Patch here, please take a look: https://github.com/include-what-you-use/include-what-you-use/pull/1491

**Comment by Jacobfaib** [comment]
> Apples libc++ should already be covered by built-in mappings in latest IWYU (but there may be omissions, of course).

I had to add the following, perhaps they are missing from the builtin rules?
```
    {
        include: ["<__fwd/sstream.h>", "private", "<sstream>", "public"]
    },
    {
        include: ["<__fwd/string_view.h>", "private", "<string_view>", "public"]
    },
    {
        include: ["<__fwd/ostream.h>", "private", "<iosfwd>", "public"]
    },
    {
        include: ["<__exception/exception.h>", "private", "<exception>", "public"]
    },
    {
        include: ["<__exception/terminate.h>", "private", "<exception>", "public"]
    },
    {
        # note, this one is not right, __hash_table is used for any of:
        # - unordered_map
        # - unordered_set
        # - set
        # - map
        include: ["<__hash_table>", "private", "<unordered_map>", "public"]
    },
    {
        include: ["<__mutex/lock_guard.h>", "private", "<mutex>", "public"]
    },
    {
        include: ["<__mutex/mutex.h>", "private", "<mutex>", "public"]
    },
```

**Comment by kimgr** [comment]
You could cross-reference with https://github.com/include-what-you-use/include-what-you-use/blob/78472a725b1dd414e0c6ccdef464a8058593d6eb/iwyu_include_picker.cc#L1006.

Looks like the `__mutex` and `__exception` ones are covered, but not the others.

**Comment by Jacobfaib** [comment]
> Patch here, please take a look: #1491

Patch LGTM!

**Comment by kimgr** [comment]
Thanks, merged in 5989280499f914c59ad28576ec64d3d9f568ede3
